crates/subspace-farmer/src/dsn/tests.rs

```diff
@@ -0,0 +1,90 @@
+use super::{sync, DSNSync, PieceIndexHashNumber, SyncOptions};
+use crate::PiecesToPlot;
+use std::{
+    collections::BTreeMap,
+    ops::Range,
+    sync::{Arc, Mutex},
+};
+use subspace_core_primitives::{Piece, PieceIndex, PieceIndexHash, PIECE_SIZE};
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+struct TestDSN(BTreeMap<PieceIndexHash, (Piece, PieceIndex)>);
+
+impl FromIterator<(Piece, PieceIndex)> for TestDSN {
+    fn from_iter<T: IntoIterator<Item = (Piece, PieceIndex)>>(iter: T) -> Self {
+        let map = iter
+            .into_iter()
+            .map(|(piece, index)| (index.into(), (piece, index)))
+            .collect();
+        Self(map)
+    }
+}
+
+#[async_trait::async_trait]
+impl DSNSync for TestDSN {
+    type Stream = futures::stream::Once<futures::future::Ready<PiecesToPlot>>;
+
+    async fn get_pieces(&mut self, Range { start, end }: Range<PieceIndexHash>) -> Self::Stream {
+        let (pieces, piece_indexes) = self
+            .0
+            .iter()
+            .skip_while(|(k, _)| **k < start)
+            .take_while(|(k, _)| **k <= end)
+            .fold(
+                (Vec::<u8>::new(), Vec::<PieceIndex>::new()),
+                |(mut flat_pieces, mut piece_indexes), (_, (piece, index))| {
+                    flat_pieces.extend(piece.iter());
+                    piece_indexes.push(*index);
+                    (flat_pieces, piece_indexes)
+                },
+            );
+        futures::stream::once(futures::future::ready(PiecesToPlot {
+            pieces: pieces.try_into().unwrap(),
+            piece_indexes,
+        }))
+    }
+}
+
+fn init() {
+    let _ = tracing_subscriber::fmt::try_init();
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn simple_test() {
+    init();
+    let dsn = (0u8..=255u8)
+        .map(|a| ([a; PIECE_SIZE].into(), a as PieceIndex))
+        .collect::<TestDSN>();
```

Comment:
You're mixing two things into one data structure unnecessarily:
* storage of sync results
* DSN sync source

Because of this it is hard to understand what is going on.

What we care about here is that we've synced what we wanted to sync:
```suggestion
    let source = (0u8..=255u8)
        .map(|a| ([a; PIECE_SIZE].into(), a as PieceIndex))
        .collect::<BTreeMap<_, _>>();
    let result = BTreeMap::new();
```
Then instantiate `TestDSN` with that `source` as `sync()` argument, but writing in the callback would be done into `result`. In the end you assert on `source` and `result` having the same contents.